crates/engine/tree/src/tree/root.rs

```diff
@@ -575,18 +575,61 @@ where
 
     /// Handles request for proof prefetch.
     fn on_prefetch_proof(&mut self, targets: MultiProofTargets) {
-        extend_multi_proof_targets_ref(&mut self.fetched_proof_targets, &targets);
+        let proof_targets = self.get_prefetch_proof_targets(targets);
+        extend_multi_proof_targets_ref(&mut self.fetched_proof_targets, &proof_targets);
 
         self.multiproof_manager.spawn_or_queue(MultiproofInput {
             config: self.config.clone(),
             hashed_state_update: Default::default(),
-            proof_targets: targets,
+            proof_targets,
             proof_sequence_number: self.proof_sequencer.next_sequence(),
             state_root_message_sender: self.tx.clone(),
             source: ProofFetchSource::Prefetch,
         });
     }
 
+    /// Calls `get_proof_targets` with existing proof targets for prefetching.
+    fn get_prefetch_proof_targets(&self, mut targets: MultiProofTargets) -> MultiProofTargets {
+        // Here we want to filter out any targets that are already fetched
+        //
+        // This means we need to remove any storage slots that have already been fetched
+        let mut duplicates = 0;
+
+        // First remove all exactly equal account / storage targets
+        targets.retain(|hashed_address, target_storage| {
+            self.fetched_proof_targets
+                .get(hashed_address)
+                .is_none_or(|fetched_storage| fetched_storage == target_storage)
```

Comment:
am I reading this wrong, or it will leave in `targets` only either accounts that do not exist in `self.fetched_proof_targets`, or accounts with storages that are equal to `self.fetched_proof_targets`?

Think it should be
```suggestion
                .is_none_or(|fetched_storage| fetched_storage != target_storage)
```

Reply:
or maybe rather it should be checking for `target_storage` being a subset of `fetched_storage`, and then removing it if so?

Reply:
Yeah, I think checking if it's a subset and removing makes more sense here